### ISSUE TITLE
Who deletes entries of BlockDirectoryCache.keysToRelease and when?

### ISSUE THREAD

**Issue by samuel281** [cloudera/lucene-solr#1]
I'm using cloudera search storing only one collection.
Documents are fed by HBase-Indexer nearly real time and there's no request to solr yet.

After reading a [documentation]
(https://www.cloudera.com/documentation/enterprise/latest/topics/search_tuning_solr.html)
I've set block cache configuration like below:
`
<directoryFactory name="DirectoryFactory">
    <bool name="solr.hdfs.blockcache.enabled">true</bool>
    <int name="solr.hdfs.blockcache.slab.count">120</int>
    <bool name="solr.hdfs.blockcache.direct.memory.allocation">true</bool>
    <int name="solr.hdfs.blockcache.blocksperbank">16384</int>
    <bool name="solr.hdfs.blockcache.read.enabled">true</bool>
    <bool name="solr.hdfs.blockcache.write.enabled">false</bool>
    <bool name="solr.hdfs.nrtcachingdirectory.enable">true</bool>
    <int name="solr.hdfs.nrtcachingdirectory.maxmergesizemb">16</int>
    <int name="solr.hdfs.nrtcachingdirectory.maxcachedmb">192</int>
</directoryFactory>
`

Couple of days later, I've found the JVM heap size kept increasing and Full GC is happening frequently. which is kinda weird cuz I'm just feeding document and there's no search query. And I set the HDFS block cache to use direct memory. Why the heap was just growing? 

So I dumped the heap and tried to figure out what's going on.
![2017-01-24 1 54 13](https://cloud.githubusercontent.com/assets/434338/22234603/b8c2ec5e-e23c-11e6-859b-33793e62e118.png)

As you can see above, there were over 10M of BlockCacheKeys and HashEntries.
If I understood the configuration correctly, the num of BlockCacheKey can't exceed 120 * 16384 (~ 1.9M). So I looked through the references and found most of HashEntries are pairs of BlockCacheKey and Boolean. And which apparently seems to be [BlockDirectoryCache.keysToRelease](https://github.com/cloudera/lucene-solr/blob/cdh5-4.4.0_5.2.0/solr/core/src/java/org/apache/solr/store/blockcache/BlockDirectoryCache.java#L46)

But I couldn't find any code that removing elements from the BlockDirectoryCache.keysToRelease.
It seems that the BlockCacheKeys are never collected by the GC. 
Is this intentional? or Is there anything I've missed?

**Comment by markrmiller** [comment]
This is a known issue that will be addressed with https://issues.apache.org/jira/browse/SOLR-9284.

**Comment by samuel281** [comment]
Hi, @markrmiller. 
I saw that ticket already closed(fixed) in v6.4 but CDH lucene-solr is still in v4. (solr-4.10.3+cdh5.10.0+463, actually)
Do you know if this could be fixed in next CDH version?
Should I just make my own patch for this?

**Comment by markrmiller** [comment]
Yeah, we backport issues all the time to the v4 version in CDH. This will be in a future release.
